### DISCUSSION TITLE
Linking  large(r) tables with the help of some code and the API [Jun 15]

### DISCUSSION THREAD

**Discussion by dstala** [nocodb/nocodb#5947] [Discord Threads]
> 👨‍🦲 **dstala**

**Comment by o1lab** [comment]
> 🙋‍♂️ **fastelbow**  
#### I finally got some time and inspiration to get this done. (there was a reason to start with NocoDB hence the no code part 😁 )   
#### I used JavaScript for it but inside a Jupyter Notebook (IjavaScript) easier tot pickup where you left and better documentation options.  
#  
> 🙋‍♂️ **fastelbow**  
#### I had to link two tables one with brand names other with model names It was a new experience for me. So I can show you how I did it and not How it should be done. I put it here in the hope it will be useful for other starters.  
#  
> 🙋‍♂️ **fastelbow**  
#### ```JS   
#### // It is a one time job I used xAuthKey replace    
#### // {xAuthKey-foundAt->Team&Settings->Team&Auth->Tab-API_Tokens_Managment} with correct token   
#### let xAuthKey = '{xAuthKey-foundAt->Team&Settings->Team&Auth->Tab-API_Tokens_Managment}'   
####    
#### // An object to give fetch the option it needs with two properties: get and post   
#### let options = {   
####   get: {method:'GET', headers: {'xc-auth': xAuthKey}},   
####   post: {method:'POST', headers: {'xc-auth': xAuthKey}}   
#### }   
####    
#### // Base name and url   
#### let baseName = 'raymus_antilles'   
#### let baseUrl = `https://nocodb.yourdomain.tst/api/v1/db/data/v1/${baseName}/`   
####    
#### // Define table names for linking   
#### let tableDic = {linkFromtbl: 'Brands', linkTotbl: 'Models'}   
####    
#### //    
#### const url = new URL(baseUrl);   
#### //------------------------------//   
#### ```  
#  
> 🙋‍♂️ **fastelbow**  
#### ```JS   
#### // A fetch request return result as JSON   
#### function fetchJSON(url, options) {   
####     return fetch(url, options).then(response => response.json())   
#### }   
####    
#### // Get the lenght of the table   
#### function getRowCount(tableName) {   
#### //     console.log(url + tableName + '/count', options.get)   
####     return fetchJSON(url + tableName + '/count', options.get)   
#### }   
####    
#### // Get the correct field name you want to link from    
#### function getBrandName(row) {   
#### //     console.log(url + tableDic.linkFromtbl + '/' + row + '?fields=AtrisMerk', options.get)   
####     return fetchJSON(url + tableDic.linkFromtbl + '/' + row + '?fields=AtrisMerk', options.get)   
#### }   
####    
#### // Collect all the Id's where you want to link-to (they need to have the same field name as in the link-from table)   
#### function getModelIds(brandName) {   
#### //     console.log(url + tableDic.linkTotbl + '?fields=Id&where=%28Merk%2Ceq%2C' + brandName + ')', options.get)   
####     return fetchJSON(url + tableDic.linkTotbl + '?fields=Id&where=%28Merk%2Ceq%2C' + brandName + ')', options.get)   
#### }   
####    
#### // Link both tables using the found Id's   
#### function linkTables(modelId, row) {   
#### //     console.log(url + tableDic.linkTotbl + '/' + modelId + '/mm/Merk%20Link/' + row, options.post)   
####     return fetch(url + tableDic.linkTotbl + '/' + modelId + '/mm/Merk%20Link/' + row, options.post)   
#### }   
#### console.log('done 1')   
#### ```  
#  
> 🙋‍♂️ **fastelbow**  
#### ```JS   
#### // Use a promise   
#### getRowCount(tableDic.linkFromtbl) // Collect the table length   
####   .then(merkCount => {   
####     // For each row in the link-from table   
####     console.log(merkCount)   
####     for (let row = 1; row <= merkCount.count; row++) {   
####         // Collect the "brand" names   
####         console.log(row)   
####         getBrandName(row)   
####             .then(brandName => {   
####             // Collect the record Id's for all link-to records you want to link to this "brand"     
####             getModelIds(brandName.AtrisMerk)   
####             .then(allModelIdsForOneBrandArr => {   
####               // Voor elk model ID in the list   
####               for (let modelId of allModelIdsForOneBrandArr.list) {   
####                 // Link de two tabellen together   
####                 linkTables(modelId.Id, row)   
####                   .then(() => console.log('Linked successfully'))   
####                   .catch(error => console.error(error))   
####               }   
####             })   
####             .catch(error => console.error(error))   
####         })   
####         .catch(error => console.error(error))   
####     }   
####   })   
####   .catch(error => console.error(error))   
#### ```  
#  
> 🙋‍♂️ **fastelbow**  
#### I had one issue with this code. In one of the brand-name fields there was a name with a ampersand "&"  and that broke the  process Or to more precise it linked almost the whole model table to that one brand. A quick workaround was to remove that one &. But that is not really a solution.    
#### Has anyone a good suggestion?   
#### And logging the real  feedback after a success full link  (or unsuccessful) would be great  
#  

 [Autogenerated from discord](https://discord.com/channels/661905455894888490/1101171749611131080)  
- - - -  
#### Join [NocoDB's](http://www.nocodb.com/)  community  
[Discord](https://discord.gg/5RgZmkW) • [Twitter](https://twitter.com/nocodb) • [Reddit](https://www.reddit.com/r/NocoDB/) • [Docs](https://docs.nocodb.com/)
